### PR TITLE
ARROW-4372: [C++] Embed precompiled bitcode in the gandiva library

### PR DESCRIPTION
**Pull request by kszucs**
We were not running the pyarrow tests after installing the manylinux wheels, which can lead to uncaught issues, like: https://travis-ci.org/kszucs/crossbow/builds/484284104

### CODE REVIEW

**Comment by kszucs** [comment]
@xhochy We're not bundling gandiva generated bytecode files. I'm trying to resolve it, but You might know the solution.

**Comment by xhochy** [comment]
> but We still need configure the engine accordingly

What do you mean with that?

**Comment by kszucs** [comment]
@xhochy according to [this commit](https://github.com/apache/arrow/commit/747590afc84481f61ead4d4c14e25ff9b79213f6#diff-37d9d708b7324782f89f277076ec0101R78) @kou made it configurable.

**Comment by kou** [comment]
The change just for creating a configuration for custom path. This is enough just for testing. But it's not suitable for normal use.
We can't change the default path yet. I think that we need a feature to customize the default path dynamically or finding the byte code path from the loaded shared library path or something.

File: .travis.yml [hunk 1]
```diff
@@ -189,6 +189,8 @@ matrix:
     - $TRAVIS_BUILD_DIR/ci/travis_script_python.sh 3.6
   - name: "[manylinux1] Python"
     language: cpp
+    env:
+    - PYTHON_VERSIONS=2.7,16 3.6,16
```
Comment (kszucs) [inline .travis.yml]: This will reduce the build times. There are nightly tests for all versions.

**Comment by kou** [comment]
> could You elaborate on customizing the default path dynamically?

The easiest way is using an environment variable:

```diff
diff --git a/cpp/src/gandiva/configuration.h b/cpp/src/gandiva/configuration.h
index 04e2eed2..fd948400 100644
--- a/cpp/src/gandiva/configuration.h
+++ b/cpp/src/gandiva/configuration.h
@@ -75,7 +75,14 @@ class ConfigurationBuilder {
   std::string byte_code_file_path_;
 
   static std::shared_ptr<Configuration> InitDefaultConfig() {
-    std::shared_ptr<Configuration> configuration(new Configuration(kByteCodeFilePath));
+    std::string path;
+    auto path_env = std::getenv("GANDIVA_BYTE_CODE_PATH");
+    if (path_env) {
+      path = path_env;
+    } else {
+      path = kByteCodeFilePath;
+    }
+    std::shared_ptr<Configuration> configuration(new Configuration(path));
     return configuration;
   }
```

But it may have a security problem.

**Comment by kszucs** [comment]
@wesm any thoughts on this?

File: ci/travis_script_manylinux.sh [hunk 2]
```diff
@@ -39,4 +44,21 @@ conda activate $CONDA_ENV_DIR
...
+popd
+
+# # Test optional dependencies and the presence of tensorflow
+python -c "
+import sys
+import pyarrow
+import pyarrow.orc
+import pyarrow.parquet
+import pyarrow.plasma
+import tensorflow
```
Comment (pitrou) [inline ci/travis_script_manylinux.sh]: Aren't we crashy when Tensorflow is loaded in the same process?

File: python/setup.py [hunk 3]
```diff
@@ -297,6 +297,10 @@ def _run_cmake(self):
                     move_shared_libs(build_prefix, build_lib, "plasma")
                 if self.with_gandiva:
                     move_shared_libs(build_prefix, build_lib, "gandiva")
+                    bytecode_files = glob.glob(pjoin(build_prefix, '*.bc'))
```
Comment (pitrou) [inline python/setup.py]: For the record, they are [_bitcode_](https://llvm.org/docs/BitCodeFormat.html) files, not _bytecode_.

File: ci/travis_script_manylinux.sh (hunk 2 above)
Comment (kszucs) [reply ci/travis_script_manylinux.sh]: I guess this is supposed to test that.

File: python/setup.py (hunk 3 above)
Comment (kszucs) [reply python/setup.py]: I'll rename :)

File: ci/travis_script_manylinux.sh (hunk 2 above)
Comment (xhochy) [reply ci/travis_script_manylinux.sh]: We should not crash, there is some special safeaguarding code for that.

Comment (pitrou) [reply ci/travis_script_manylinux.sh]: But it doesn't always work, see https://issues.apache.org/jira/browse/ARROW-4272

File: python/setup.py (hunk 3 above)
Comment (kou) [reply python/setup.py]: Oh, we need to rename "bytecode" in Gandiva. @pravindra

Comment (pravindra) [reply python/setup.py]: I guess bc can mean both :-). I'll fix any references to byte code in gandiva code.

**Comment by kszucs** [comment]
@pravindra couldn't We embed the bitcode into the library?

**Comment by pravindra** [comment]
> @pravindra couldn't We embed the bitcode into the library?

sorry, I didn't understand. can you please elaborate some more ? The bit-code file is generated by a series of clang tool commands during the build process. Is there a way to add such a file into the .so or .a, and extract it later at runtime ?

**Comment by pitrou** [comment]
The easiest and most portable way (though slightly clunky) would probably to generate a C++ file containing static `const char` arrays of the embedded bitcodes...

See https://stackoverflow.com/questions/11813271/embed-resources-eg-shader-code-images-into-executable-library-with-cmake for an example, though we'd probably want to write the generator utility in Python.

**Comment by kszucs** [comment]
@pravindra yes, like Antoine's example. Although it might be easier to just configure the right path from the python module (similarly like from java). Not sure which one would be cleaner...

**Comment by kszucs** [comment]
But We still need to keep the ability to load bitecode from files, right?

**Comment by pitrou** [comment]
Apparently the bitcode path is hardcoded as an absolute path currently (see the generated `bc_file_path.cc`), so the approach would have to be fixed anyway:
```cpp
namespace gandiva {

// Path to the byte-code file.
extern const char kByteCodeFilePath[] = "/home/antoine/miniconda3/envs/pyarrow/lib/gandiva/irhelpers.bc";

} // namespace gandiva
```

**Comment by kszucs** [comment]
So We have a couple of options:
- don't set default path at all, and let the binding (java, python) configure it 
- set default path relative to the shared lib (is it fragile in case of static libs?)
- compile the bitcode as a string into the lib

which one should We go with?

**Comment by pitrou** [comment]
I think compiling the bitcode as a string is the most pain-free approach. Other approaches have to deal with locating the files on disk in different environments (Python, Java, ...).

**Comment by kszucs** [comment]
Any downstream libraries would like to provide precompiled gandiva expressions should do the same?

**Comment by pitrou** [comment]
I think that's a separate concern. It would probably be possible to let them choose their preferred way.

**Comment by kszucs** [comment]
Ok, I'll go with the string approach then. Thanks @pitrou !

**Comment by pravindra** [comment]
I do see the advantages of the embed-as-string approach.

Early on, we had plans to support pluggable function registries in gandiva I.e given a new bitcode file and the fn signatures, it would be possible  to add new functions to gandiva by doing a module reload  (without actually changing arrow/gandiva code). so, passing the location(s) of the bitcode file(s) as a config param made sense.

We didn’t get around to implementing it properly, and I think it’s harder to do now since there are cross references from the bitcode files into gandiva/arrow (for date/decimal). 

Embedding the bc code into the library will mean that we cannot go down that path anymore.

**Comment by wesm** [comment]
Where do things stand on this?

**Comment by kszucs** [comment]
I can continue to work on this tomorrow, right now I'm busy with buildbot's infra.

File: cpp/src/gandiva/projector_cache_key.h [hunk 4]
```diff
@@ -41,7 +41,7 @@ class ProjectorCacheKey {
       boost::hash_combine(result, expr_as_string);
       UpdateUniqifier(expr_as_string);
     }
-    boost::hash_combine(result, configuration->Hash());
+    // boost::hash_combine(result, configuration->Hash());
```
Comment (kszucs) [inline cpp/src/gandiva/projector_cache_key.h]: I'm not sure that We want to hash the precompiled bitcode, so I've disabled it.

File: cpp/src/gandiva/tests/projector_test.cc [hunk 5]
```diff
@@ -88,17 +88,17 @@ TEST_F(TestProjector, TestProjectCache) {
...
+  // auto pos = alt_path.find('/', 0);
+  // EXPECT_NE(pos, std::string::npos);
+  // alt_path.replace(pos, 1, "//");
+  // auto other_configuration =
+  //     ConfigurationBuilder().set_byte_code_file_path(alt_path).build();
+  // std::shared_ptr<Projector> should_be_new_projector2;
+  // status = Projector::Make(schema, {sum_expr, sub_expr}, other_configuration,
+  //                          &should_be_new_projector2);
+  // ASSERT_OK(status);
+  // EXPECT_NE(projector, should_be_new_projector2);
```
Comment (kszucs) [inline cpp/src/gandiva/tests/projector_test.cc]: We don't hash byte_code_file_path anymore, so I plan to remove it.

File: ci/travis_script_manylinux.sh [hunk 6]
```diff
@@ -21,14 +21,18 @@
 set -ex
 
 pushd python/manylinux1
-docker run --shm-size=2g --rm -e PYARROW_PARALLEL=3 -v $PWD:/io -v $PWD/../../:/arrow quay.io/xhochy/arrow_manylinux1_x86_64_base:llvm-7-manylinux1 /io/build_arrow.sh
+docker run --shm-size=2g --rm \
+  -e PYARROW_PARALLEL=3 \
+  -v $PWD:/io \
+  -v $PWD/../../:/arrow \
+  quay.io/xhochy/arrow_manylinux1_x86_64_base:llvm-7-manylinux1 \
```
Comment (kszucs) [inline ci/travis_script_manylinux.sh]: Ideally We should not run the test inside the image, just outside to prevent the issues like this.

Comment (xhochy) [reply ci/travis_script_manylinux.sh]: Yes, running them somewhere else shound ensure that we have packaged all necessary things into the wheel.

File: cpp/src/gandiva/configuration.h [hunk 7]
```diff
@@ -38,17 +38,17 @@ class GANDIVA_EXPORT Configuration {
...
   bool operator!=(const Configuration& other) const;
 
  private:
-  explicit Configuration(const std::string& byte_code_file_path)
-      : byte_code_file_path_(byte_code_file_path) {}
+  explicit Configuration(const std::string& precompiled_bitcode)
+      : precompiled_bitcode_(precompiled_bitcode) {}
 
-  const std::string byte_code_file_path_;
+  const std::string& precompiled_bitcode_;
```
Comment (pravindra) [inline cpp/src/gandiva/configuration.h]: This is not likely to be useful. Let's remove the precompiled_bitcode_ completely ?

File: cpp/src/gandiva/jni/config_builder.cc [hunk 8]
```diff
@@ -34,16 +34,16 @@ using gandiva::ConfigurationBuilder;
...
-  jstring byte_code_file_path =
+  jstring precompiled_bitcode =
       (jstring)env->CallObjectMethod(configuration, byte_code_accessor_method_id_, 0);
   ConfigurationBuilder configuration_builder;
-  if (byte_code_file_path != nullptr) {
-    const char* byte_code_file_path_cpp = env->GetStringUTFChars(byte_code_file_path, 0);
-    configuration_builder.set_byte_code_file_path(byte_code_file_path_cpp);
-    env->ReleaseStringUTFChars(byte_code_file_path, byte_code_file_path_cpp);
+  if (precompiled_bitcode != nullptr) {
+    const char* precompiled_bitcode_cpp = env->GetStringUTFChars(precompiled_bitcode, 0);
```
Comment (pravindra) [inline cpp/src/gandiva/jni/config_builder.cc]: I think passing the entire byte-code buffer across ini for every project/filter call is too much overhead. let's remove this (always use the cc file) ?

File: cpp/src/gandiva/projector_cache_key.h (hunk 4 above)
Comment (pravindra) [reply cpp/src/gandiva/projector_cache_key.h]: maybe, keep this Hash(), but dummy out the configuration->Hash() (just return 0) ?

File: cpp/src/gandiva/tests/projector_test.cc (hunk 5 above)
Comment (pravindra) [reply cpp/src/gandiva/tests/projector_test.cc]: Is it ok if I give you a patch for the test ?

File: cpp/src/gandiva/configuration.h (hunk 7 above)
Comment (kszucs) [reply cpp/src/gandiva/configuration.h]: I've removed it, much better now :) 
I was hesitating to remove it, wasn't sure about the plans.

File: cpp/src/gandiva/jni/config_builder.cc (hunk 8 above)
Comment (kszucs) [reply cpp/src/gandiva/jni/config_builder.cc]: Agree, removed.

**Comment by kszucs** [comment]
@pravindra bundling works properly now, but it cannot load `libgandiva_jni` and it's quiet hard to get a meaningful stack trace out of mvn.

**Comment by pravindra** [comment]
thanks @kszucs. I'll fix the jni part & send you a patch (hopefully today).

**Comment by kszucs** [comment]
@pravindra what's the issue BTW? It really bugs me :)

File: .travis.yml (hunk 1 above)
Comment (kszucs) [reply .travis.yml]: @pitrou This will resolve the random timeouts of wheel builds.

**Comment by kszucs** [comment]
It works for me now locally.

**Comment by pravindra** [comment]
> It works for me now locally.

thanks for fixing this. nice catch !

File: ci/travis_script_manylinux.sh (hunk 2 above)
Comment (pitrou) [reply ci/travis_script_manylinux.sh]: So shouldn't we remove this import?

File: cpp/src/gandiva/engine.cc [hunk 9]
```diff
@@ -66,6 +66,10 @@
 
 namespace gandiva {
 
+GANDIVA_EXPORT
```
Comment (pitrou) [inline cpp/src/gandiva/engine.cc]: Why the export? I'm expecting the bitcode object file to be linked statically with this, and never referenced from the outside.

File: cpp/src/gandiva/precompiled/CMakeLists.txt [hunk 10]
```diff
@@ -67,12 +67,32 @@ foreach(SRC_FILE ${PRECOMPILED_SRCS})
...
                            ${BC_FILES}
                    DEPENDS ${BC_FILES})
 
-add_custom_target(precompiled ALL DEPENDS ${GANDIVA_BC_OUTPUT_PATH})
+# write a cmake script to replace precompiled bitcode file's content into a .cc file
+file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/WritePrecompiledCC.cmake" "\
+  file(READ \"${GANDIVA_PRECOMPILED_BC_PATH}\" HEXCONTENT HEX)
+  string(REGEX REPLACE
+         \"([0-9a-f][0-9a-f])\"
+         \"'\\\\\\\\x\\\\1',\"
```
Comment (pitrou) [inline cpp/src/gandiva/precompiled/CMakeLists.txt]: Wow.

**Comment by pitrou** [comment]
Thanks for doing this @kszucs !

File: cpp/src/gandiva/engine.cc (hunk 9 above)
Comment (kszucs) [reply cpp/src/gandiva/engine.cc]: It was used for the bitcode path previously. Removing.

File: cpp/src/gandiva/precompiled/CMakeLists.txt (hunk 10 above)
Comment (kszucs) [reply cpp/src/gandiva/precompiled/CMakeLists.txt]: Don't even mention :)

File: ci/travis_script_manylinux.sh (hunk 2 above)
Comment (kszucs) [reply ci/travis_script_manylinux.sh]: Why should We? If We want to test that both arrow and tensorflow can live in the same interperter.

Comment (pitrou) [reply ci/travis_script_manylinux.sh]: As you prefer... but it will probably fail at one point or the other, because Tensorflow publishes rogue wheels.

**Comment by wesm** [comment]
Oh the things we do for software =)

**Comment by kszucs** [comment]
We need to ensure that the remaining wheels (OSX and Windows) and conda packages are shipping gandiva as well.

File: dev/tasks/python-wheels/travis.linux.yml [hunk 11]
```diff
@@ -30,7 +30,7 @@ env:
     - TRAVIS_TAG={{ task.tag }}
 
 before_script:
-  - docker pull quay.io/xhochy/arrow_manylinux1_x86_64_base:latest
```
Comment (xhochy) [inline dev/tasks/python-wheels/travis.linux.yml]: This should be latest again.

File: python/manylinux1/build_arrow.sh [hunk 12]
```diff
@@ -48,95 +45,73 @@ export PYARROW_CMAKE_OPTIONS='-DTHRIFT_HOME=/usr -DBoost_NAMESPACE=arrow_boost -
...
+    --bundle-boost \
+    --boost-namespace=arrow_boost
+PATH="$PATH:${CPYTHON_PATH}/bin" $PYTHON_INTERPRETER setup.py bdist_wheel
+PATH="$PATH:${CPYTHON_PATH}/bin" $PYTHON_INTERPRETER setup.py sdist
+
+echo "=== (${PYTHON_VERSION}) Tag the wheel with manylinux1 ==="
+mkdir -p repaired_wheels/
+auditwheel -v repair -L . dist/pyarrow-*.whl -w repaired_wheels/
+
+# Testing happens outsite of the build to prevent issues like ARROW-4372
```
Comment (xhochy) [inline python/manylinux1/build_arrow.sh]: Can you keep the import checks here? They are fast and already provide good insights.

File: cpp/Dockerfile [hunk 13]
```diff
@@ -48,6 +48,7 @@ RUN arrow/ci/docker_install_conda.sh && \
 
 ENV CC=gcc \
     CXX=g++ \
+    ARROW_GANDIVA=OFF \
```
Comment (xhochy) [inline cpp/Dockerfile]: Why is this disabled?

File: python/manylinux1/build_arrow.sh (hunk 12 above)
Comment (kszucs) [reply python/manylinux1/build_arrow.sh]: Good point!

**Comment by kszucs** [comment]
Follow-up issues:
- ARROW-4644: [C++/Docker] Build Gandiva in the docker containers
- ARROW-4645: [C++/Packaging] Ship Gandiva with OSX and Windows wheels
- ARROW-4646: [C++/Packaging] Ship gandiva with the conda-forge packages

File: cpp/Dockerfile (hunk 13 above)
Comment (kszucs) [reply cpp/Dockerfile]: Created a follow-up issue.

File: python/manylinux1/build_arrow.sh (hunk 12 above)
Comment (kszucs) [reply python/manylinux1/build_arrow.sh]: Added.

File: dev/tasks/python-wheels/travis.linux.yml (hunk 11 above)
Comment (kszucs) [reply dev/tasks/python-wheels/travis.linux.yml]: Done.

**Review by pravindra — APPROVED** [review]
the gandiva changes LGTM. I skipped over the python changes.

**Comment by kszucs** [comment]
Thanks for your review @pravindra!

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**